src/invariant/InvariantTests.cpp

```diff
@@ -39,9 +53,494 @@ class TestInvariant : public Invariant
+    LedgerDelta ld(lh, app->getDatabase(), false);
+
+    uint32_t stopLedgerSeq = ledgerSeq + nLedgers;
+    for ( ; ledgerSeq <= stopLedgerSeq; ++ledgerSeq)
+    {
+        auto live = LedgerTestUtils::generateValidLedgerEntries(5);
+        for (auto& entry : live)
+        {
+            entry.lastModifiedLedgerSeq = ledgerSeq;
+            EntryFrame::FromXDR(entry)->storeAdd(ld, app->getDatabase());
+        }
+        std::vector<LedgerKey> dead;
+
+        auto& bl = app->getBucketManager().getBucketList();
+        bl.addBatch(*app, ledgerSeq, live, dead);
+    }
+    return ledgerSeq;
+}
+
+void
+getHistoryArchiveState(Application::pointer appGenerate,
+                       Application::pointer appApply,
+                       uint32_t ledgerSeq,
+                       std::map<std::string, std::shared_ptr<Bucket>>& buckets,
+                       HistoryArchiveState& has)
+{
+    auto& blGenerate = appGenerate->getBucketManager().getBucketList();
+    auto& bmApply = appApply->getBucketManager();
+
+    for (size_t i = 0; i <= BucketList::kNumLevels - 1; i++)
+    {
+        auto& level = blGenerate.getLevel(i);
+        auto curr = level.getCurr();
+        auto snap = level.getSnap();
+        buckets[binToHex(curr->getHash())] =
+            bmApply.adoptFileAsBucket(curr->getFilename(), curr->getHash());
+        buckets[binToHex(snap->getHash())] =
+            bmApply.adoptFileAsBucket(snap->getFilename(), snap->getHash());
+    }
+    has = HistoryArchiveState(ledgerSeq, blGenerate);
+}
+
+template <typename T = ApplyBucketsWork, typename... Args>
+void
+applyBucketsAndCrankUntilDone(Application::pointer appGenerate,
+                              Application::pointer appApply,
+                              uint32_t ledgerSeq,
+                              Args&&... args)
+{
+    std::map<std::string, std::shared_ptr<Bucket>> buckets;
+    HistoryArchiveState has;
+    getHistoryArchiveState(appGenerate, appApply, ledgerSeq, buckets, has);
+
+    auto& wm = appApply->getWorkManager();
+    wm.addWork<T>(buckets, has, std::forward<Args>(args)...);
+    wm.advanceChildren();
+    auto& clock = appApply->getClock();
+    while (!clock.getIOService().stopped() && !wm.allChildrenDone())
+    {
+        clock.crank(false);
+    }
+}
+
+std::string
+getTableName(LedgerEntryType t)
+{
+    switch (t)
+    {
+    case ACCOUNT:
+        return "accounts";
+    case TRUSTLINE:
+        return "trustlines";
+    case OFFER:
+        return "offers";
+    case DATA:
+        return "accountdata";
+    default:
+        REQUIRE(false);
+        return {};
+    }
+}
+
+template <typename T>
+void
+bucketListConsistent(
+        const std::function<std::vector<uint32_t>(Database&,
+                                                  LedgerEntryType,
+                                                  uint32_t)>& f)
+{
+    auto seed = std::chrono::system_clock::now().time_since_epoch().count();
+    std::default_random_engine gen(seed);
```

Comment:
forgot to mention: do not mess with the seed in test cases to keep test runs deterministic. If you need more randomness, iterate more.